### PR TITLE
Added basic management operations.

### PR DESCRIPTION
**Pull request by Maximelego**
Added database Initialization.
Added Redis manager for tokens.
Added Basic routers.
Added basic enumerations for HTTP Exceptions.

All models are subject to change about their data. The only useful ones are login and hash.
Added basic authentication methods.
Updated docker containers. **You will need to rebuild the project !**

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -1,14 +1,58 @@
...
-app = FastAPI()
+from app.routes import auth, user
 
-@app.get("/")
-async def root():
-    return RedirectResponse(url="/docs")
+from app.utils.databases.db import startup_databases
+
+# Array for the routes descriptions and names.
+tags_metadata = [
```
Comment (desaleo) [inline app/main.py:10]: should be typed with list[dict[str, str]] or using TypedDict

File: app/main.py [hunk 2]
```diff
@@ -1,14 +1,58 @@
...
+        "description": "User operations."
+    },
+    {
+        "name": "auth",
+        "description": "Authentication endpoints."
+    }
+]
+
+# Creation of the main router
+app = FastAPI(title="SOBEK W.A. API",
```
Comment (desaleo) [inline app/main.py:22]: should be type with FastAPI object

File: app/main.py [hunk 3]
```diff
@@ -1,14 +1,58 @@
...
+              description="API For the SOBEK W.A web application",
+              openapi_tags=tags_metadata)
+
+# List of requests origins that are allowed for the API
+# IMPORTANT NOTICE: Using the wildcard * is dangerous:
+# It will allow all request sources.
+# Furthermore, with authentication enabled, it will not work.
+# Listing authorized sources is MANDATORY.
+origins = [
```
Comment (desaleo) [inline app/main.py:31]: should be typed with list[str]

File: app/main.py [hunk 4]
```diff
@@ -1,14 +1,58 @@
...
+    allow_headers=["*"],
+)
 
+# Importing API routes :
+app.include_router(user.userRouter, prefix="/user", tags=["user"])
+app.include_router(auth.authRouter, prefix="/auth", tags=["auth"])
 
+# TODO : Fix the deprecated event.
+@app.on_event("startup")
+async def startup():
```
Comment (desaleo) [inline app/main.py:51]: should be typed (return type: None)

File: app/main.py [hunk 5]
```diff
@@ -1,14 +1,58 @@
...
+    await startup_databases(app=app)
+
+
+# Root path: Redirecting to the documentation.
+@app.get("/")
+async def root():
```
Comment (desaleo) [inline app/main.py:57]: should be typed

File: app/models/Tokens.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+from fastapi import HTTPException
+from passlib.context import CryptContext
+
+import enum
+from typing import Final
+
+from app.utils.databases.redis import Redis
+from app.utils.http_errors import CommonErrorMessages, ClassicExceptions
+
+pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")
```
Comment (desaleo) [inline app/models/Tokens.py:17]: should be typed

File: app/models/Tokens.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+# -------- Classic models -------- #
+class Token:
+    value: str | None = None
+    attributes: TokenAttributes
+
+    def __init__(self, attributes: TokenAttributes, value: str | None = None):
+        self.attributes = attributes
+        self.value      = value
+
+    def generate(self, user_id: int):
```
Comment (desaleo) [inline app/models/Tokens.py:64]: should be typed (return type: None)

File: app/models/Tokens.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        expire_date = datetime.now() + timedelta(minutes=float(self.attributes.expire_time))
+        jwt_data = {
+            "user_id": user_id,
+            "salt": str(bcrypt.gensalt()),
+            "iat": creation_date,
+            "exp": expire_date
+        }
+        self.value = jwt.encode(jwt_data, str(self.attributes.secret), self.attributes.algorithm)
+
+    async def revoke(self, redis_db: redis.Redis = Redis.get_redis()):
```
Comment (desaleo) [inline app/models/Tokens.py:81]: should be typed with Awaitable[\<ReturnType>] or Coroutine[\<YieldType>, \<SendType>, \<ReturnType>]

File: app/models/Tokens.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            return
+
+        # Extracting payload
+        data = await self.extract_payload()
+        # Adding the token to redis blocklist with an expiration date according to the expiration date of the token.
+        redis_db.setex(self.value, data["exp"], self.attributes.token_type)
+        self.value = None
+
+
+    async def extract_payload(self):
```
Comment (desaleo) [inline app/models/Tokens.py:95]: should be typed with Awaitable[<ReturnType>] or Coroutine[<YieldType>, <SendType>, <ReturnType>]

File: app/models/Tokens.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+        except jwt.exceptions.InvalidTokenError:
+            # This exception is raised if the token cannot be decoded.
+            raise HTTPException(status_code=401, detail=CommonErrorMessages.TOKEN_INVALID)
+
+        # We return the payload
+        return data
+
+
+    async def is_revoked(self, redis_db: redis.Redis = Redis.get_redis()):
```
Comment (desaleo) [inline app/models/Tokens.py:125]: should be typed with Awaitable[<ReturnType>] or Coroutine[<YieldType>, <SendType>, <ReturnType>]

File: app/models/Tokens.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+class TokenPair:
+    access_token:  Token
+    refresh_token: Token
+
+    def __init__(self, access_token: str | None = None, refresh_token: str | None = None):
+        self.access_token  = Token(AvailableTokenAttributes.AUTH_TOKEN.value, access_token)
+        self.refresh_token = Token(AvailableTokenAttributes.REFRESH_TOKEN.value, refresh_token)
+
+
+    def get_tokens_in_response(self):
```
Comment (desaleo) [inline app/models/Tokens.py:143]: should be typed using dict[] or TypedDict

File: app/models/Tokens.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+
+    def revoke_tokens(self) -> None:
+        """
+        This method revokes the access token and refresh token inside the current object.
+        """
+        self.access_token.revoke()
+        self.refresh_token.revoke()
+
+
+    async def refresh_tokens(self) -> None:
```
Comment (desaleo) [inline app/models/Tokens.py:171]: should be typed with Awaitable[<ReturnType>] or Coroutine[<YieldType>, <SendType>, <ReturnType>]

File: app/models/pydantic/TokenModel.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+from app.models.Tokens import TokenAttributes, Token, TokenPair
+
+# -------- pydantic models -------- #
+class PydanticTokenExporter(ABC, BaseModel):
+    """
+    This class provides an abstract method that forces us to define an abstract method that exports
+    the current pydantic model to a Tortoise one.
+    """
+    @abstractmethod
+    def export_pydantic_to_model(self, attributes: TokenAttributes | None = None):
```
Comment (desaleo) [inline app/models/pydantic/TokenModel.py:18]: should be typed (return type)

File: app/models/pydantic/TokenModel.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+        """
+        This method exports the current pydantic model
+        """
+        pass
+
+class PydanticToken(PydanticTokenExporter):
+    value: str
+    def export_pydantic_to_model(self, attributes: TokenAttributes | None = None):
```
Comment (desaleo) [inline app/models/pydantic/TokenModel.py:26]: should be typed (return type)

File: app/models/pydantic/TokenModel.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+        return Token(attributes=attributes, value=self.value)
+
+class PydanticTokenPair(PydanticTokenExporter):
+    access_token: str
+    refresh_token: str
+    def export_pydantic_to_model(self, attributes: TokenAttributes | None = None):
```
Comment (desaleo) [inline app/models/pydantic/TokenModel.py:32]: should be typed (return type)

File: app/models/tortoise/user.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+from fastapi import Depends, HTTPException
+from fastapi.security import OAuth2PasswordBearer
+from passlib.context import CryptContext
+from tortoise import fields
+from tortoise.models import Model
+
+from app.models.pydantic.TokenModel import PydanticToken
+from app.utils.http_errors import ClassicExceptions
+
+pwd_context   = CryptContext(schemes=["bcrypt"], deprecated="auto")
```
Comment (desaleo) [inline app/models/tortoise/user.py:17]: should be typed and the line below as well

File: app/models/tortoise/user.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    It contains all the columns necessary to store the user.
+    """
+    id = fields.IntField(primary_key=True)
+    login = fields.CharField(unique=True, required=True, max_length=128)
+    firstname = fields.CharField(max_length=128)
+    lastname = fields.CharField(max_length=128)
+    mail = fields.CharField(unique=True, required=True, max_length=128)
+    hash = fields.TextField(required=True)
+
+    def __str__(self):
```
Comment (desaleo) [inline app/models/tortoise/user.py:33]: should be typed (return type: str)

File: app/models/Tokens.py [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+    secret:      str
+
+    def __init__(self, token_type: TokenTypes):
+        self.token_type = token_type                                # <- Type of the token. Used to get specific env vars.
+
+        load_dotenv()
+        self.algorithm  = os.environ.get("JWT_ALGORITHM", "HS256")  # <- Algorithm used for encoding the tokens
+        self.expire_time = os.environ.get(token_type.value + "_EXPIRE", None)
+        if self.expire_time is None:
+            raise Exception("No expire time was provided inside the .env file !")
```
Comment (desaleo) [inline app/models/Tokens.py:40]: should use a more precise exception class, even better: a custom one

File: app/models/Tokens.py [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+
+        self.secret = os.environ.get("JWT_SECRET_" + token_type.value, None)
+        if self.expire_time is None:
+            raise Exception("No secrets were provided inside the .env file !")
```
Comment (desaleo) [inline app/models/Tokens.py:44]: should use a more precise exception class, even better: a custom one

File: app/models/tortoise/user.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+    def get_password_hash(password) -> str :
+        """
+        This Method is designed to get the Hash of a password
+        """
+        password_hash = bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt())
+        return password_hash.decode('utf-8')
+
+
+    @staticmethod
+    async def authenticate_user(login: str, password: str) -> Optional["UserInDB"] :
```
Comment (desaleo) [inline app/models/tortoise/user.py:49]: should be typed with Awaitable[<ReturnType>] or Coroutine[<YieldType>, <SendType>, <ReturnType>]

File: app/models/tortoise/user.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+    firstname = fields.CharField(max_length=128)
+    lastname = fields.CharField(max_length=128)
+    mail = fields.CharField(unique=True, required=True, max_length=128)
+    hash = fields.TextField(required=True)
+
+    def __str__(self):
+        """
+        This method will output a str describing the UserInDB class.
+        """
+        return "[INFO] - USER " + self.login + " ID : " + str(self.id)
```
Comment (desaleo) [inline app/models/tortoise/user.py:37]: should use f-string instead like `` f"[INFO] - USER {self.login} ID : {self.id}"``

File: app/routes/auth.py [hunk 22]
```diff
@@ -0,0 +1,3 @@
+from fastapi import APIRouter
+
+authRouter = APIRouter()
```
Comment (desaleo) [inline app/routes/auth.py:3]: should be typed

File: app/routes/user.py [hunk 23]
```diff
@@ -0,0 +1,3 @@
+from fastapi import APIRouter
+
+userRouter = APIRouter()
```
Comment (desaleo) [inline app/routes/user.py:3]: should be typed

File: app/utils/databases/datasets.py [hunk 24]
```diff
@@ -0,0 +1,2 @@
+async def load_dummy_datasets():
```
Comment (desaleo) [inline app/utils/databases/datasets.py:1]: should be typed

File: app/utils/databases/db.py [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+
+from dotenv import load_dotenv
+from fastapi import FastAPI
+
+from app.utils.databases.datasets import load_dummy_datasets
+from app.utils.databases.postgresql import Postgresql
+from app.utils.databases.redis import Redis
+
+
+async def startup_databases(app: FastAPI):
```
Comment (desaleo) [inline app/utils/databases/db.py:11]: should be typed

File: app/utils/databases/redis.py [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+import redis
+from dotenv import load_dotenv
+
+
+class Redis:
+
+    redis = None
+
+    @classmethod
+    async def load_redis(cls):
```
Comment (desaleo) [inline app/utils/databases/redis.py:12]: should be typed (return type: None)

File: app/utils/databases/postgresql.py [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+        """
+        This method returns a list of available models.
+        These are located inside the tortoise module.
+        """
+        model_files = list(
+            filter(lambda x : x.endswith(".py") and x != "__init__.py", os.listdir("./app/models/tortoise")))
+        return map(lambda x : "app.models.tortoise." + x[:-3], model_files)
+
+    @staticmethod
+    async def init_postgres_db(app: FastAPI) -> None:
```
Comment (desaleo) [inline app/utils/databases/postgresql.py:49]: should be typed with Awaitable[<ReturnType>] or Coroutine[<YieldType>, <SendType>, <ReturnType>]

File: app/utils/databases/postgresql.py [hunk 28]
```diff
@@ -0,0 +1,74 @@
...
+                elements_missing.append("user ")
+            if password is None :
+                elements_missing.append("password ")
+            raise Exception("Some parameters are missing : " + str(elements_missing))
+
+        return f"{dms}://{user}:{password}@{host}:{port}/{db}"
+
+
+    @staticmethod
+    def get_available_models():
```
Comment (desaleo) [inline app/utils/databases/postgresql.py:39]: should be typed

File: app/utils/databases/redis.py [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+        load_dotenv(".env")
+        pool = redis.ConnectionPool(host=os.environ.get("REDIS_HOST"),
+                                    port=os.environ.get("REDIS_PORT"),
+                                    db=os.environ.get("REDIS_DB"),
+                                    password=os.environ.get("REDIS_PASSWORD"))
+
+        cls.redis = redis.Redis(connection_pool=pool)
+
+    @classmethod
+    def get_redis(cls):
```
Comment (desaleo) [inline app/utils/databases/redis.py:22]: should be typed

**Review by desaleo — CHANGES_REQUESTED** [review]
A lot of un-typed/miss-typed values (I've commented every one i've seen but I may have missed some especially in function body) I highly recommand using pyright as a type checker in strict mode.

**Review by HugoBien — APPROVED** [review]
Great work !

**Review by desaleo — APPROVED** [review]
LGTM

**Pull request merged**